### ISSUE TITLE
Can't add first class callable filter: No closure found

### ISSUE THREAD

**Issue by spaze** [efabrica-team/phpstan-latte#435]
When adding a filter like:
```php
$engine->addFilter('objectFilterFirstClassCallable', (new Filters())->objectFilter(...));
```
PHPStan returns "No closure found". I don't even know where to start so I'll just add a failing test (#436).

**Comment by MartinMystikJonas** [comment]
This looks like limitation of roave/betterreflection which is used to analyse closure created by first class callable. It tries to analyse given closure (param and return types) by parsing known location (file+line) of closure definition (it is part of closure "object") but looks only for normal closures not for first class callables.

Honestly I have no idea how to fix this.

**Comment by MartinMystikJonas** [comment]
I will check is there is any workaround for this used in PHPStan itself somewhere.

**Comment by MartinMystikJonas** [comment]
Reported issue to roave/betterreflection https://github.com/Roave/BetterReflection/issues/1363

**Comment by ondrejmirtes** [comment]
Hey, please let me know the place in code where this problem occurs for you.

Personally I think if you did `$scope->getType()` instead of going through BetterReflection you should be able to avoid this problem.

**Comment by MartinMystikJonas** [comment]
It happens here: https://github.com/efabrica-team/phpstan-latte/blob/af52e65f0d7c5bfd29e4625ebc1f3449f0a2a660/src/Compiler/NodeVisitor/ChangeFiltersNodeVisitor.php#L239


We have Closure (real) created as first class callable which is extracted from Latte engine. And we need to know its signature to be able to preprocess code for analysis. Replace magic call by direct call to given method for which first class callable was created.

**Comment by ondrejmirtes** [comment]
> we need to know its signature to be able to preprocess code for analysis

In which form do you need the signature? What about FunctionReflection/MethodReflection or ParametersAcceptor?

**Comment by MartinMystikJonas** [comment]
Well, class name + method name of first class callablectarget would be enough. Or we need types of all parameters and return type to create matching placeholder - MethodReflection should be enough.

Problem is we do not have scope from plave where closure is defined - we just get closure itself from latte engine.

**Comment by ondrejmirtes** [comment]
Why is this code in a node visitor and not done later where you Scope and other helpers available?

**Comment by MartinMystikJonas** [comment]
We compile template by latte, use these preprocessors to make it analysable and save it. Then we run analysis of this compiled template.

**Comment by ondrejmirtes** [comment]
Please show me these steps:

1) Original latte template
2) Latte template after transformation analysable by PHPStan (where the aforementioned filter is used and the AST is changed by this visitor)

**Comment by MartinMystikJonas** [comment]
Original call to filter looks like this (`$this->filters` is jsut universal crate for filters, template have no idea what is in it at runtime)
```php
($this->filters->upper)($title)
```
for static methods we transform it to:
```php
\Nette\Utils\Strings::upper($title);
```
for non-static methods we use:
```php
/** callable(string):string $__filter_upper */
// ...
$__filter_upper($title)
```

**Comment by ondrejmirtes** [comment]
Why don't you transform the filter call to:

```php
((new Filters())->objectFilter(...))($whatIFilter)
```

That way you don't really need to care and transform the details of the call.

**Comment by ondrejmirtes** [comment]
Or maybe just do:

```php
$__filter_upper = The\Callable::doFoo(...);
```

**Comment by ondrejmirtes** [comment]
Otherwise you could also do the job of collecting "this filter has this ParametersAcceptor signature" with a custom Collector (where you have all the PHPStan internals available) and then use that information later.

**Comment by MartinMystikJonas** [comment]
Problem is we have only Closure object we get from latte engine. It is created outside analysed files.

**Comment by MartinMystikJonas** [comment]
For your alternatives - in both cases we need to thow class name and method name Closure targets. But we cannot get it from Closure itslef.

**Comment by MartinMystikJonas** [comment]
We also need to know if first param of target is FilterInfo because it is added to call magically by filter executor outside template

**Comment by ondrejmirtes** [comment]
I get it - so you don't see the AST from `$engine->addFilter('objectFilterFirstClassCallable', (new Filters())->objectFilter(...));`, you're trying to understand it and recreate from a `Closure` object.

The bug isn't in BetterReflection. You're calling `ReflectionFunction::createFromClosure` which is only for functions anyway, not methods.

Since you already have the `Closure`, I think you can just read the info manually by doing `new \ReflectionFunction($closure)` in the visitor, and read it your way. That should always work. And should be faster too.

**Comment by MartinMystikJonas** [comment]
Thanks for the tip, I will try it.

**Comment by MartinMystikJonas** [comment]
Fixed by #450 

@ondrejmirtes Thanks for good tip
